src/server/failed_logins.rs

```diff
@@ -0,0 +1,188 @@
+use crate::options::FailedLoginsPenalty;
+use crate::options::FailedLoginsPolicy;
+
+use super::shutdown;
+use slog::Logger;
+use std::collections::HashMap;
+use std::net::IpAddr;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+use tokio::sync::{Mutex, RwLock};
+
+#[derive(Hash, Eq, PartialEq, Debug, Clone)]
+struct FailedLoginsKey {
+    ip: Option<IpAddr>,
+    username: Option<String>,
+}
+
+#[derive(Debug, Clone)]
+struct FailedLoginsEntry {
+    attempts: u32,
+    last_attempt_at: Instant,
+}
+
+impl FailedLoginsEntry {
+    fn new() -> Mutex<FailedLoginsEntry> {
+        Mutex::new(FailedLoginsEntry {
+            attempts: 1,
+            last_attempt_at: Instant::now(),
+        })
+    }
+
+    fn time_elapsed(&self) -> Duration {
+        self.last_attempt_at.elapsed()
+    }
+
+    fn touch(&mut self) {
+        self.last_attempt_at = Instant::now();
+    }
+}
+
+/// Temporarily remembers failed logins
+#[derive(Debug)]
+pub struct FailedLoginsCache {
+    policy: FailedLoginsPolicy,
+    penalty: FailedLoginsPenalty,
+    failed_logins: Arc<RwLock<HashMap<FailedLoginsKey, Mutex<FailedLoginsEntry>>>>,
+}
+
+#[derive(Debug)]
+pub enum FailedLoginsError {
+    // With this failed login attempt the lockout threshold has been reached
+    MaxFailuresReached,
+    // The account is already locked out from previous failed attempts
+    AlreadyLocked,
+}
+
+impl FailedLoginsCache {
+    pub fn new(failed_logins_policy: FailedLoginsPolicy) -> Arc<FailedLoginsCache> {
+        let penalty = match failed_logins_policy {
+            FailedLoginsPolicy::SourceIPLock(ref x) => x.clone(),
+            FailedLoginsPolicy::SourceUserLock(ref x) => x.clone(),
+            FailedLoginsPolicy::UserLock(ref x) => x.clone(),
+        };
+
+        Arc::new(FailedLoginsCache {
+            policy: failed_logins_policy,
+            penalty,
+            failed_logins: Arc::new(RwLock::new(HashMap::new())),
+        })
+    }
+
+    fn is_expired(&self, time_elapsed: Duration) -> bool {
+        time_elapsed > self.penalty.expires_after
+    }
+
+    fn is_locked(&self, attempts: u32) -> bool {
+        attempts >= self.penalty.max_attempts
+    }
+
+    fn getkey(&self, ip: IpAddr, user: String) -> FailedLoginsKey {
+        match self.policy {
+            FailedLoginsPolicy::SourceUserLock(_) => FailedLoginsKey {
+                ip: Some(ip),
+                username: Some(user),
+            },
+            FailedLoginsPolicy::SourceIPLock(_) => FailedLoginsKey { ip: Some(ip), username: None },
+            FailedLoginsPolicy::UserLock(_) => FailedLoginsKey {
+                ip: None,
+                username: Some(user),
+            },
+        }
+    }
+
+    /// Upon failed login: increments failed attempts counter, returns error if account is locked out
+    pub async fn failed(&self, ip: IpAddr, user: String) -> Result<(), FailedLoginsError> {
```

Comment:
The alternative to having `FailedLoginsError` implement the Error trait is to not see this as an exceptional case and rather just return an option e.g.

`pub async fn failed(&self, ip: IpAddr, user: String) -> Option<LockState> {`

where LockState = : 

```
pub enum LockState {
   // With this failed login attempt the lockout threshold has been reached
    MaxFailuresReached,
    // The account is already locked out from previous failed attempts
    AlreadyLocked,
}
```

Reply:
Yes, definitely. I think at some point that nagged in the back of my mind but I was focused on the functionality. Thanks for pointing this out! Will fix.